### PR TITLE
misc: add method to parse dictionary attr MLIR style to BaseParser

### PR DESCRIPTION
**Pull request by superlopuh**
Currently calls an MLIRParser method with an instance of XDSLParser, seems dangerous.

### CODE REVIEW

File: xdsl/parser.py [hunk 1]
```diff
@@ -1464,6 +1464,25 @@ def try_parse_builtin_arr_attr(self) -> ArrayAttr | None:
     def parse_optional_attr_dict(self) -> dict[str, Attribute]:
         raise NotImplementedError()
 
+    def parse_optional_dictionary_attr_dict(self) -> dict[str, Attribute]:
+        if not self.tokenizer.starts_with("{"):
+            return dict()
+
+        self.parse_characters(
+            "{",
+            "MLIR Attribute dictionary must be enclosed in curly brackets")
```
Comment (webmiche) [inline xdsl/parser.py:1473]: This error message is out of sync now, right? This does not just apply to the MLIR one, but to general ones.

Comment (superlopuh) [reply xdsl/parser.py:1473]: Yep

**Review by AntonLydike — APPROVED** [review]
LGTM

**Review by math-fehr — APPROVED** [review]
Nice! Thanks

**Pull request merged**